[Testing] Add EVM E2E load test using Vegeta

commoddity opened the pull request:
## 🌿 Summary

Add EVM E2E load test using Vegeta for comprehensive performance testing of EVM JSON-RPC methods.

### ⛓ Services Being Tested in the E2E

**Morse**
- `F00C` (Ethereum)
- `F021` (Polygon)
- `F01C` (Oasys)
- `F036` (XRPL EVM Testnet)

**Shannon**
- `anvil`

### 🌱 Primary Changes:
- Implement new EVM load test framework using Vegeta for performance testing
- Add support for testing 9 core EVM JSON-RPC methods with configurable parameters
- Replace old Morse/Shannon relay tests with new EVM-focused test structure
- Added E2E tests docs page

### 🍃 Secondary changes:
- Add progress bars and detailed logging for test visualization
- Improve Docker test setup with force rebuild and logging controls
- Add helper functions for building JSON-RPC params for different method types

## 🛠️ Type of change

Select one or more from the following:

- [x] New feature, functionality or library

## 🤯 Sanity Checklist

- [x] I have updated the GitHub Issue 'assignees', 'reviewers', 'labels', 'project', 'iteration' and 'milestone'
- [x] For code, I have run 'make test_all'
- [x] I added TODOs where applicable

Olshansk reviewed (dismissed):
@commoddity Requesting changes because it doesn't seem the large merge fonclifct was handled correctly: 

<img width="788" alt="Screenshot 2025-04-16 at 2 02 07 PM" src="https://github.com/user-attachments/assets/34d8b6ad-e2ca-4c13-ab4f-5b4ef2917a22" />

Please  re-request a review once its ready.

commoddity commented:
> @commoddity Requesting changes because it doesn't seem the large merge fonclifct was handled correctly:
> 
> <img alt="Screenshot 2025-04-16 at 2 02 07 PM" width="788" src="https://private-user-images.githubusercontent.com/1892194/434528115-34d8b6ad-e2ca-4c13-ab4f-5b4ef2917a22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5OTA5NjIsIm5iZiI6MTc0NDk5MDY2MiwicGF0aCI6Ii8xODkyMTk0LzQzNDUyODExNS0zNGQ4YjZhZC1lMmNhLTRjMTMtYWI0Zi01YjRlZjI5MTdhMjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MThUMTUzNzQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2QzNzk1ZmNiNTk3ZDlkMjY3N2IwMjE3M2JkY2MwZjY4NjM3MGU3ZDMwYzFkZjU4MzdhZTYzYzA1NTBjODA1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.eExemxjOGG7gEuCR8U8eujf4SpcodO7VlPKafVWkltw">
> Please re-request a review once its ready.

My mistake. I have updated the PR to fix these merge conflict markers.

Also added a docs page about the E2E tests. PTAL when you can @Olshansk .

adshmh commented on e2e/evm_methods_test.go:
```diff
@@ -0,0 +1,247 @@
+
+// methodDefinitions contains all method definitions for a service load test.
+// this allows customizing the configuration for each method as desired.
+var methodDefinitions = map[jsonrpc.Method]methodDefinition{
```
Not a blocker, but this looks like a good candidate for reading the settings from a yaml file.

commoddity replied:
Sure that could be a good future feature but not sure that it adds a lot of value currently, what about adding it as a TODO_FUTURE or TODO_IMPROVE?

adshmh commented on e2e/vegeta_test.go:
```diff
@@ -0,0 +1,824 @@
+}
+
+// runAttack executes a load test for the given method
+func runAttack(
```
I recommend a more descriptive name here.

commoddity replied:
The term"attack" comes from the Vegeta library, which specifically calls sending a request an attack. WDYT about maybe just adding a comment that adds this context?

adshmh commented on observation/protocol/morse.pb.go:
```diff
@@ -31,8 +31,10 @@ const (
 	MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_INVALID_RESPONSE                    MorseEndpointErrorType = 6
-	MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_NON_JSON_RESPONSE                   MorseEndpointErrorType = 7 // Response returned from endpoint is not valid JSON
-	MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_INTERNAL                            MorseEndpointErrorType = 8 // Added for internal gateway errors
+	// This error indicates that the endpoint returned invalid data in its response and was rejected by the SDK.
```
These should be added to the .proto file, or they will be overwritten each time we regenerate `.pb.go` files.

commoddity replied:
I've confirmed that it is added to the .proto file.

adshmh commented on qos/jsonrpc/request.go:
```diff
@@ -53,16 +53,34 @@ func (r Request) MarshalJSON() ([]byte, error) {
+
+// BuildParamsFromStringArray builds a Params object from an array of strings.
 //
-// JSON-RPC array params must be passed in the order specified by the method.
```
This message looks helpful, is it being delete because it is inaccurate?

adshmh reviewed (approved):
The PR looks good to me, just a few non-blockers to consider.
